test/integration.sh

```diff
@@ -1,47 +1,93 @@
 #!/usr/bin/env bash
-set -e
+
+set -e  # exit if return is non-zero
+# Start logging.
 source $(dirname $( realpath ${BASH_SOURCE[0]} ) )/lib/b-log.sh
 
 function check_setup {
-    [[ $(sudo docker ps | tail -n +2 | wc -l) -ge 7 ]] || (echo "the required docker containers don't seem to be running" && exit 1)
-    # TODO count number of dashboard containers
-    # TODO more rigorous checking
-}
 
-function main {
-    check_setup
-    cd "$(dirname "${BASH_SOURCE[0]}")"
-    LOG_LEVEL_ALL
+    # This function exits the script if:
+    #   (i) the number of running Docker is less than
+    #       the required number of containers
+    #   (ii) not all exact names of required Docker containers
+    #        match all names in the list of running Docker containers
+
+    # Get array of running containers.
+    containers=($(sudo docker ps | tail -n +2 | awk ' {print $NF} '))
+
+    # Get array of all required containers from file.
+    fname=expected_containers
+    IFS=$'\r\n' GLOBIGNORE='*'\
+       command eval  'containers_expected=($(cat $fname))'
+    containers_expected=($containers_expected)  # cast as array
 
-    [[ -d outputs ]] && sudo rm -rf outputs && INFO "deleted old outputs"
+    num_containers_exp="${#containers_expected[@]}"
 
-    access_token=$(../redwood/cli/bin/redwood token create)
-    INFO "generated testing access token: ${access_token}"
+    # Crude initial test to verify that at least the minimum number of
+    # containers are running.
+    [ "${#containers[@]}" -ge $num_containers_exp ] ||\
+        (ERROR "Not running the required number of docker containers."\
+                && exit 1)
 
-    redwood_endpoint=$(cat ../redwood/.env | grep 'base_url=' | sed 's/[^=]*=//')
-    INFO "got redwood endpoint from dcc-ops/redwood/.env: ${redwood_endpoint}"
+    for container_expected in "${containers_expected[@]}"; do
+        counter=0
+        for container in "${containers[@]}"; do
+            if [[ $container_expected = $container ]]; then
+                continue
+            else
+                counter=$((counter+1))
+            fi
+        done
 
-    INFO "doing upload with $(pwd)/manifest.tsv"
-    sudo docker run --rm -it -e ACCESS_TOKEN=${access_token} -e REDWOOD_ENDPOINT=${redwood_endpoint} \
-         -v $(pwd)/manifest.tsv:/dcc/manifest.tsv -v $(pwd)/samples:/samples -v $(pwd)/outputs:/outputs \
-         quay.io/ucsc_cgl/core-client:1.1.0-alpha spinnaker-upload --skip-submit --force-upload /dcc/manifest.tsv
+        if [[ "$counter" -eq $num_containers_exp ]]; then
+            ERROR "One or more docker containers not running."\
```

Comment:
I did not look at this loop closely the first time... I don't think it's worth the effort to spend more time on this ticket, because things will either work or they won't and so there's no point in overkill, but.. since you had the container name that wasn't found, you could have displayed the missing container name rather than a generic message.